File: Java/src/_1818_Minimum_Absolute_Sum_Difference/Solution.java
package _1818_Minimum_Absolute_Sum_Difference;

import java.util.TreeSet;

/**
 * Date: 2021/7/14 10:11
 * Content:
 */
public class Solution {

    static final int mod = (int) (1e9) + 7;

    public int minAbsoluteSumDiff(int[] nums1, int[] nums2) {
        int n = nums1.length;
        TreeSet<Integer> a = new TreeSet<>();
        for (int x : nums1) a.add(x);
        int sum = 0, d = 0;
        for (int i = 0; i < n; i++) {
            int t = Math.abs(nums1[i] - nums2[i]);
            sum = (sum + t) % mod;
            if (a.ceiling(nums2[i]) != null)
                d = Math.max(d, t - Math.abs(a.ceiling(nums2[i]) - nums2[i]));
            if (a.floor(nums2[i]) != null)
                d = Math.max(d, t - Math.abs(a.floor(nums2[i]) - nums2[i]));
        }
        return (sum - d + mod) % mod;
    }

    public static void main(String[] args) {

        int[] nums1 = {1, 28, 21};
        int[] nums2 = {9, 21, 20};
        System.out.println((new Solution()).minAbsoluteSumDiff(nums1, nums2));  // 9
    }
}

File: Python/_1818_Minimum_Absolute_Sum_Difference/Solution.py
# coding=utf-8
# Date: 2021/7/14 10:19
from typing import List


# 执行用时：860 ms, 在所有 Python3 提交中击败了35.67%的用户
# 内存消耗：26.9 MB, 在所有 Python3 提交中击败了69.94%的用户
class Solution:
    def minAbsoluteSumDiff(self, nums1: List[int], nums2: List[int]) -> int:
        mod = int(1e9) + 7
        n = len(nums1)
        a = nums1.copy()
        a.sort()
        sum, d = 0, 0
        for i in range(n):
            t = abs(nums1[i] - nums2[i])
            sum = (sum + t) % mod
            l, r = 0, n
            while l < r:
                mid = l + r >> 1
                if a[mid] >= nums2[i]:
                    r = mid
                else:
                    l = mid + 1
            if r < n:
                d = max(d, t - abs(a[r] - nums2[i]))
            if r > 0:
                d = max(d, t - abs(a[r - 1] - nums2[i]))
        return (sum - d + mod) % mod


if __name__ == "__main__":
    print(Solution().minAbsoluteSumDiff([1, 28, 21], [9, 21, 20]))  # 9
